### PR TITLE
[hwasan] Allow stack traces even when fixed shadow is used

### PR DESCRIPTION
**Pull request by thurstond**
Previously, fixed shadow implied !InTls, and !InTls implied no stack traces, but InTls implied it would not use a fixed shadow. This patch changes fixed shadow to be compatible with stack traces.

It maintains the legacy behavior for KHWAsan || InstrumentWithCalls.

### CODE REVIEW

File: llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp [hunk 1]
```diff
@@ -1897,16 +1897,19 @@ void HWAddressSanitizer::ShadowMapping::init(Triple &TargetTriple,
...
-  } else if (ClMappingOffset.getNumOccurrences() > 0) {
-    InGlobal = false;
-    InTls = false;
-    Offset = ClMappingOffset;
-    WithFrameRecord = false;
   } else if (ClEnableKhwasan || InstrumentWithCalls) {
     InGlobal = false;
     InTls = false;
-    Offset = 0;
+    if (ClMappingOffset.getNumOccurrences() > 0)
```
Comment (fmayer) [inline llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:1904]: `optOr`.

**Review by fmayer — COMMENTED** [review]
advance copy of nit, will think about the PR tomorrow

File: llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp [hunk 2]
```diff
@@ -1340,7 +1340,8 @@ Value *HWAddressSanitizer::getFrameRecordInfo(IRBuilder<> &IRB) {
 }
 
 void HWAddressSanitizer::emitPrologue(IRBuilder<> &IRB, bool WithFrameRecord) {
-  if (!Mapping.InTls)
+  if (!Mapping.InTls ||
+      (Mapping.Offset != 0 && Mapping.Offset != kDynamicShadowSentinel))
```
Comment (fmayer) [inline llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:1344]: this is essentially `isFixedShadow`, right? maybe pull this into a variable of that name for documentation.

**Review by fmayer — COMMENTED** [review]
Tests are failing

File: llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp [hunk 3]
```diff
@@ -1897,16 +1898,16 @@ void HWAddressSanitizer::ShadowMapping::init(Triple &TargetTriple,
...
-  } else if (ClMappingOffset.getNumOccurrences() > 0) {
-    InGlobal = false;
-    InTls = false;
-    Offset = ClMappingOffset;
-    WithFrameRecord = false;
   } else if (ClEnableKhwasan || InstrumentWithCalls) {
     InGlobal = false;
     InTls = false;
-    Offset = 0;
+    Offset = optOr<uint64_t>(ClMappingOffset, (uint64_t)0);
```
Comment (fmayer) [inline llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:1904]: Why is this necessary? ClMappingOffset is opt<uint64_t> so I don't understand why we need the explicit specialization.

Also please don't use C-style casts.

**Comment by vitalybuka** [comment]
Alternative set of patches:
#109616
#109617
#109618
#109619
#109620
#109621

File: llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp (hunk 2 above)
Comment (vitalybuka) [reply llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:1344]: https://github.com/llvm/llvm-project/pull/109616
